src/morphological_operations.jl

```diff
@@ -99,3 +99,73 @@ which is defined as the arithmetic difference between the internal and the exter
 `region` allows you to control the dimensions over which this operation is performed.
 """
 morpholaplace(img::AbstractArray, region=coords_spatial(img)) = dilate(img, region) + erode(img, region) - 2img
+
+
+"""
+```
+function thinning_iteration(img::AbstractArray{Bool,2}; odd_iteration::Bool)
+```
+The thining iteration evaluates three conditions in order to determine which pixels of the image should be removed from img.
+The three conditions are explained in:
+Guo, Z., & Hall, R. W. (1989). Parallel thinning with two-subiteration algorithms. Communications of the ACM, 32(3), 359-373.
+"""
+function thinning_iteration!(img_ori::AbstractArray{Bool,2}, odd_iteration::Bool)
+    marker = trues(size(img_ori))
+    img = falses(size(img_ori).+2)
+    img[2:end-1,2:end-1]=img_ori
+    h, w = size(img)
+    for i=2:h-1
+        for j=2:w-1
+            if !img[i,j]
+                continue
+            end
+            p1 = img[i-1,j-1]
+            p2 = img[i-1,j]
+            p3 = img[i-1,j+1]
+            p4 = img[i,j+1]
+            p5 = img[i+1,j+1]
+            p6 = img[i+1,j]
+            p7 = img[i+1,j-1]
+            p8 = img[i,j-1]
+            A = (!p2 && (p3 || p4)) + (!p4 && (p5 || p6)) + (!p6 && (p7 || p8)) + (!p8 && (p1 || p2))
+            B1 = (p1 || p2) + (p3 || p4) + (p5 || p6) + (p7 || p8)
+            B2 = (p2 || p3) + (p4 || p5) + (p6 || p7) + (p8 || p1)
+            B = min(B1,B2)
+            G3 = false
+            if (odd_iteration)
+                G3 = (p2 || p3 || (!p5)) && p4
+            else
+                G3 = (p6 || p7 || (!p1)) && p8
+            end
+            if (A == 1) && ((2 <= B) && (B <= 3)) && (!G3) 
+                marker[i-1,j-1] =false
+            end
+        end
+        
+    end
+    img_ori[:] = img_ori[:].&(marker[:])
+end
+
+doc"""
+```julia
+function thinning(img::AbstractArray{Bool}; maxiter::Integer=0) 
+```
+
+Applies a binary blob thinning operation, to achieve a skeletization of the input image.
+The procedure carries on until max_iterations is achieved or when there are no changes between two consecutives iterations.
+
+The algorithm is described in:
+1. Guo, Z., & Hall, R. W. (1989). Parallel thinning with two-subiteration algorithms. Communications of the ACM, 32(3), 359-373.
+2. Lam, L., Lee, S. W., & Suen, C. Y. (1992). Thinning methodologies-a comprehensive survey. IEEE Transactions on pattern analysis and machine intelligence, 14(9), 869-885.
+"""
+function thinning(img::AbstractArray{Bool}; maxiter::Integer=0) 
+    prev = falses(size(img))
+    processed = copy(img)
+    it = 0
+    while ((maxiter==0) || (it<maxiter)) && (sum(prev.!=processed)>0)
```

Comment:
Yes, the iterations are idempotent after a few steps. I wrote it not because a matter of convergence, but in case that is useful for someone not to perform the iterations until convergence. I remember I read some methods that used some threshold to stop the procedure. But sincecerly I couldn't find references about that. So, I suppose is not such a very common usecase.

Reply:
Makes sense, I thought you had this intention in mind :) Is there a situation where we want to thin an image to an intermediate (not fully thinned) state? I am just curious of the applications, it should be fine to keep the parameter there.